src/main/java/org/apache/brooklyn/cloudfoundry/entity/VanillaCloudfoundryApplicationImpl.java

```diff
@@ -117,7 +152,28 @@ protected void preStart(Collection<? extends Location> locations) {
         }
     }
 
-    //Probably it would be better an Optional object
+    /*
+     * TODO: avoiding boilerplate code
+     * This method is a copy of getLocations in MachineLifecycleEffectorTasks
+     */
+    protected Location findLocation(@Nullable Collection<? extends Location> locations) {
+        if (locations == null || locations.isEmpty()) {
+            locations = this.getLocations();
+        }
+
+        locations = Locations.getLocationsCheckingAncestors(locations, this);
+
+        Maybe<MachineLocation> ml = Locations.findUniqueMachineLocation(locations);
+        if (ml.isPresent()) return ml.get();
+
+        if (locations.isEmpty())
```

Comment:
Yes, it can be empty. Indeed I have added a test case to check the exception throwing